edgelet/edgelet-docker/src/runtime.rs

```diff
@@ -829,6 +778,84 @@ where
     )
 }
 
+fn authenticate<MR>(
+    runtime: &MR,
+    req: &Request<Body>,
+) -> Box<dyn Future<Item = AuthId, Error = Error> + Send>
+where
+    MR: ModuleRuntime<Error = Error>,
+    <MR as ModuleRuntime>::ListFuture: 'static,
+    MR::Module: DockerModuleTop<Error = Error>,
+{
+    let pid = req
+        .extensions()
+        .get::<Pid>()
+        .cloned()
+        .unwrap_or_else(|| Pid::None);
+
+    let expected_module_id = req.extensions().get::<ModuleId>().cloned();
+
+    Box::new(match pid {
+        Pid::None => Either::A(future::ok(AuthId::None)),
+        Pid::Any => Either::A(future::ok(AuthId::Any)),
+        Pid::Value(pid) => {
+            Either::B(
+                // to authenticate request we need to determine whether given pid corresponds to
+                // any pid from any module. In order to do so, we are getting a list of all running
+                // modules and for each of them execute docker top command. There can be errors
+                // during a multiple requests, so we are filtered out those modules that we active
+                // during docker inspect operation but have gone after (NotFound and TopModule errors).
+                runtime
+                    .list()
+                    .map(move |list| {
+                        stream::futures_unordered(list.into_iter().filter_map(|module| {
+                            if expected_module_id.is_none()
+                                || *expected_module_id.as_ref().unwrap() == module.name()
+                            {
+                                Some(module.top())
+                            } else {
+                                None
+                            }
+                        }))
+                    })
+                    .into_stream()
```

Comment:
After you change the previous `map` to filter `list` synchronously, you also won't need this `.into_stream().flatten().then()` any more.